### PR TITLE
vicdual: nsub gradient

### PR DESCRIPTION
**Pull request by SailorSat**
I figured out how the starfield / gradient get enabled on vicdual hardware.

N-Sub uses a gradient for the background; there are black destroyers on the horizon that basically can't be seen without.

N-Sub and Star Raker also seem to have some kind of starfield simulation which is completly missing for now.

Can't find any decent schematics for the hardware, so the gradient is based on guesswork for now.

Reference for N-Sub: https://www.youtube.com/watch?v=Lqq4ZKyh3UA

(older) simuation (column based) seemed off at the water surface - http://images.arianchen.de/mame/nsub/0003.png

(current) simulation (pixel based) is better, but still not perfect -  http://images.arianchen.de/mame/nsub/0004.png

### CODE REVIEW

**Comment by ajrhacker** [comment]
All the available Vic Dual schematics indicate that the palette select uses only two bits, though the same word is also sent to an extra "Sound Out" connector that doesn't seem to be typically used.

Space Trek is another Vic Dual game that should feature a starfield and a blue gradient, though with the blue color fading in rather than out. This gradient can be switched on or off; however, the schematics for this game are unavailable.
https://www.youtube.com/watch?v=xsEP_oUygzY

The starfields are probably controlled by the timing PROMs, which MAME currently doesn't use.

File: src/mame/drivers/polyplay.cpp [hunk 1]
```diff
@@ -262,7 +262,7 @@ ADDRESS_MAP_END
 
 static ADDRESS_MAP_START( polyplay_io_zrepp, AS_IO, 8, polyplay_state )
 	AM_IMPORT_FROM(polyplay_io_zre)
-	// TODO: add SIO ports here
+	AM_RANGE(0x88, 0x8b) AM_DEVREADWRITE(Z80SIO_TAG, z80ctc_device, read, write)
```
Comment (startaq) [inline src/mame/drivers/polyplay.cpp]: `z80ctc_device`?

Comment (SailorSat) [inline src/mame/drivers/polyplay.cpp]: oops :)

**Comment by SailorSat** [comment]
It seems there might be up to 4 bits used, but only the lower two are used to switch between palettes.
Based on the schematics I got:

OUTPUT + A2 trigger U18 (LS175)

DO 0 -> 13 = 15 & !14 -> COLOR PROM A3
DO 1 -> 12 = 10 & !11 -> COLOR PROM A4
DO 2 -> 5  = 7  & !6
DO 3 -> 4  = 2  & !3

It appears these 4 outputs are available on a connector G, together with +5V and GND.

**Comment by SailorSat** [comment]
Some line comes from U83 (PAL) but sadly that part of the schematics are hard to read.
Anyone has a better copy  of the schematics than the one on archive.org?

Some line comes from U83 (PAL) but sadly that part of the schematics are hard to read.
Anyone has a better copy  of the schematics than the one on archive.org?

The color PROM is labeled U11 (82S123, 32x8 PROM); 4 bits for foreground, 4 bits for background.
These 8 bits get fet into U9 and U10.

The RGBX output itself is done by U8 (LS157) which is driven by the VIDEO signal. This is basically the foreground/background switch. The palette data are fed by U9 and U10, which are triggered by the PAL.

My first idea was, the PAL is used to fiddle with this - like switching between foreground and background color. This proved to be a dead end, as the foreground color on the gradient depends on what object was drawn there last. Hence it can be anything from yellow, to magenta, to white.

**Comment by SailorSat** [comment]
(I really need to check by board...)
The wiring diagram I got shows no additional hardware besides the sound board and the credit unit.

As the VIDEO signal, as well as the palette bits 2 an 3 are output on some connectors, I strongly believe there is an additional board installed on the games that use gradients/starfields that is not shown in the schematics.

This board would use h- and v-sync to trigger some circuit that uses the VIDEO bit and the additional palette bits to manipulate the RGBX signal

**Comment by SailorSat** [comment]
As I have both a carnival and n-sub pcb available I should be able to cross check for additional hardware.

**Comment by ajrhacker** [comment]
The MAME driver mentions an extra 97269-P-B PCB for N-Sub, which is not included in the schematics.

**Comment by SailorSat** [comment]
Oh dear... It even states there is another PROM on it...

**Comment by SailorSat** [comment]
This is gonna be fun...
Boardset: http://images.arianchen.de/2017-01/030320171774.jpg
Addon-Board: http://images.arianchen.de/2017-01/030320171775.jpg

I traced all of the lines to various locations on the board (detail will follow).
I need to desolder the wires (and most likely the PROM) to get decent images.

**Comment by Osso13** [comment]
z80sio_device doesn't have "read" and "write" methods. This breaks compile.

**Comment by SailorSat** [comment]
Oh dear... Sorry.
Must've been braindead lately.

I'll provide a proper fix.

**Pull request merged**